game/menu/gamemanager.gd

```diff
@@ -13,22 +20,162 @@ func _ready():
 
 
 func load_game(game_cfg: ConfigFile):
+	_last_loaded_game = game_cfg.get_meta("folder_name")
+	_started_playing_game = OS.get_ticks_msec()
 	# load the games main scene
-	var scene = load(game_cfg.get_meta("folder_path") + game_cfg.get_value("game", "main_scene"))
-
+	var scene = load(
+		(
+			"res://games/%s/%s"
+			% [
+				game_cfg.get_meta("folder_name"),
+				game_cfg.get_value("game", "main_scene"),
+			]
+		)
+	)
 	var err := get_tree().change_scene_to(scene)
 	if err != OK:
 		prints("Error", err)
 		return
 	_main.hide()
 
 
+## Save data to a file with for the given game.
+## This is a private function, don't use this in a game
+func _save_data(file_name: String, data, game: String) -> int:
+	var cache_key = get_current_player() + "/" + game + "/" + file_name
+	var directory = "user://" + get_current_player() + "/" + game + "/"
+	Directory.new().make_dir_recursive(directory)
+	if not cache_key in GAME_DATA_CACHE:
+		GAME_DATA_CACHE[cache_key] = _load_data(file_name, game)
+	GAME_DATA_CACHE[cache_key] = data
+	var file = File.new()
+	file.open(directory + file_name + ".json", File.WRITE)
+	file.store_string(JSON.print(GAME_DATA_CACHE[cache_key]))
+	file.close()
+	return OK
+
+
+## Load data from a file or GAME_DATA_CACHE and return the data for the game.
+## If something doesn't exist an empty Dictionary is returned, which is put in
+## the correct location in the GAME_DATA_CACHE Dictionary.
+## This is a private function, don't use this in a game.
+func _load_data(file_name: String, game: String) -> Dictionary:
+	var cache_key = get_current_player() + "/" + game + "/" + file_name
+	if cache_key in GAME_DATA_CACHE:
+		return GAME_DATA_CACHE[cache_key]
+
+	var directory = "user://" + get_current_player() + "/" + game + "/"
+	Directory.new().make_dir_recursive(directory)
+
+	GAME_DATA_CACHE[cache_key] = {}  # populate with default
+	# read saved data from a file
+	var file = File.new()
+	file.open(directory + file_name + ".json", File.READ)
+
+	if not file.is_open():
+		return GAME_DATA_CACHE[cache_key]
+
+	var content = file.get_as_text()
+	file.close()
+
+	var parse_result = JSON.parse(content)
+	if not parse_result.error:
+		# put contents from file in the cache
+		GAME_DATA_CACHE[cache_key] = parse_result.result
+
+	return GAME_DATA_CACHE[cache_key]
+
+
+## Save the changes to the dict returned by get_game_data()
+## This method is automatically called when a game ends.
+func save_game_data():
+	assert(_last_loaded_game != null)  # this should only be called if _last_loaded_game is set
+	_save_data("game_data", _load_data("game_data", _last_loaded_game), _last_loaded_game)
```

Comment:
Isn't this already called whenever something is saved?
Maybe use a dirtybit instead to see what parts of the GAME_DATA_CACHE changed and save only those?

Reply:
This doesn't save the whole GAME_DATA_CACHE. This only saved the game_data for the current game.
The games can get a Dictionary which they can modify, the changes they do to this Dictionary will get saved